IMPROVE - Add test coverage for OdometerViewModel

lauriecai opened the pull request:
# What it Does
Resolves #379 by validating CRUD operations for `OdometerViewModel`
* Tests initial state, add/delete/get/update odometer readings, and get vehicles

# How I Tested
* Set `useEmulator` to `true` within `OdometerViewModelTests` `init` function
* Initialize each test with a `userUID`, `viewModel`, and `newReading`
* Test CRUD operations by adding, deleting, updating, and fetching `newReading`, then checking database for `newReading`

# Notes
* This is my first contribution in a codebase that isn't my own and is this large - would appreciate any feedback with regard to approach, things to consider, alternatives, folder structure, etc.
* Is issue the same as a PR? I created a new issue #379, but this ticket contains more information about what was done. Is this redundant or meant to be separate? Is there a better way to link the two?

lauriecai committed:
Create FirebaseService to handle vehicle fetch and CRUD operations for odometer readings

lauriecai committed:
Refactor OdometerViewModel to utilize FirebaseService

lauriecai committed:
Add FirebaseServiceProtocol for mocking Firebase interactions in tests

lauriecai committed:
Add MockFirebaseService for tests

lauriecai committed:
Add test coverage for OdometerViewModel

lauriecai committed:
Utilize protocol-based dependency injection for Firebase service

mikaelacaron commented:
Thank you for contributing!

The point of the issue is because it's more of the idea of "what are you planning to do" it generally doesn't have the exact specifics of how to complete that issue

The PR is more about describing what exactly you did and why. (Because it's closer related to the git history and will have details that may be helpful if there's a bug found in the future related to this PR)

So how you wrote them is fine! I'm about to review the PR itself, so I may make more notes, but generally for this project you should be using the Firebase emulator. Is there anything in the [CONTRIBUTING instructions](https://github.com/mikaelacaron/Basic-Car-Maintenance/blob/dev/CONTRIBUTING.md) that I can add that makes it easier to understand what it's for and why you should be using it when contributing to this project?

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/ViewModels/OdometerViewModel.swift:
```diff
@@ -24,21 +24,20 @@ class OdometerViewModel {
+    func addReading(_ reading: OdometerReading) throws {
         if let uid = userUID {
-            var readingToAdd = odometerReading
+            var readingToAdd = reading
```
be careful of little changes like this. changing `odometerReading` to `reading` isn't necessary and adds more lines of code changed than is required

lauriecai commented:
Thanks for the explanation - I'll do some more investigation on Firebase emulator and how to write better tests around that (instructions were clear, I was just worried about biting off more than I could chew). Will come back with a proposed plan. If you do end up with some feedback before that, please share!

mikaelacaron commented on Basic-Car-Maintenance/Shared/Utilities/FirebaseService.swift:
```diff
@@ -0,0 +1,82 @@
+
+class FirebaseService: FirebaseServiceProtocol {
+    
+    static let shared = FirebaseService()
+    
+    private init() { }
+    
```
can we do this without making `FirebaseService` a singleton? Because I generally want to avoid singletons

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/Views/OdometerView.swift:
```diff
@@ -186,67 +186,55 @@ enum TimeRange: String, CaseIterable, Identifiable {
-    let fifthReading = createReading(vehicleID: secondCar.id!,
-                                      date: "2024/10/1",
-                                      distance: 1000)
+    let readings = [
+        createReading(vehicleID: firstCar.id!, date: "2024/10/18", distance: 35),
+        createReading(vehicleID: firstCar.id!, date: "2024/10/19", distance: 564),
```
this is another small change that doesn't actually need to happen

Instead of making individual variables, now it's changing a bunch of lines of code just because, but the code itself isn't really changed, this is purely stylistic

lauriecai replied:
Having some trouble with this one. When I revert back to the original code, it gives this error:
`Cannot use explicit 'return' statement in the body of result builder 'ViewBuilder'`

for the following two blocks:
```
return OdometerView(viewModel: viewModel)
        .environment(ActionService.shared)`
```
```
return OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: firstDate, 
                               distance: distance, 
                               isMetric: false, 
                               vehicleID: vehicleID)
```

I recall this working at one point, but not sure what's causing this issue now.

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/Views/OdometerView.swift:
```diff
@@ -186,67 +186,55 @@ enum TimeRange: String, CaseIterable, Identifiable {
+        .environment(ActionService.shared)
+}
+
+func createMockViewModel() -> OdometerViewModel {
```
these functions should live inside the `#Preview` and not just floating at the bottom of the file

* createMockViewModel
* createVehicle
* createReading

lauriecai replied:
Having some trouble with this one. When I revert back to the original code, it gives this error:

`Cannot use explicit 'return' statement in the body of result builder 'ViewBuilder'`

for the following two blocks:
```
return OdometerView(viewModel: viewModel)
        .environment(ActionService.shared)`
```
```
return OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: firstDate, 
                               distance: distance, 
                               isMetric: false, 
                               vehicleID: vehicleID)
```
I recall this working at one point, but not sure what's causing this issue now.

lauriecai replied:
Was able to resolve the error and keep everything contained within `#Preview` by removing helper functions `createVehicle()` and `createReading()`, but it's not as clean compared to what you previously had. I might need some help with this. Are you seeing the same error with the original code?

```
#Preview {
    let viewModel = OdometerViewModel(userUID: nil, firebaseService: FirebaseService())
    
    viewModel.vehicles.append(contentsOf: [
        Vehicle(id: "id1", 
                userID: nil, 
                name: "My 1st car", 
                make: "", 
                model: "", 
                year: nil, 
                color: nil, 
                vin: nil, 
                licensePlateNumber: nil),
        Vehicle(id: "id2", 
                userID: nil, 
                name: "2nd Car", 
                make: "", 
                model: "", 
                year: nil, 
                color: nil, 
                vin: nil, 
                licensePlateNumber: nil)
    ])
    
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy/MM/dd"
    
    viewModel.readings.append(contentsOf: [
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/10/18")!, 
                               distance: 35, 
                               isMetric: false, 
                               vehicleID: "id1"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/10/19")!, 
                               distance: 564, 
                               isMetric: false, 
                               vehicleID: "id1"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/11/23")!, 
                               distance: 1000, 
                               isMetric: false, 
                               vehicleID: "id1"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/11/30")!, 
                               distance: 1024, 
                               isMetric: false, 
                               vehicleID: "id1"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/10/1")!, 
                               distance: 1000, 
                               isMetric: false, 
                               vehicleID: "id2"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/10/13")!, 
                               distance: 1144, 
                               isMetric: false, 
                               vehicleID: "id2"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/10/15")!, 
                               distance: 1412, 
                               isMetric: false, 
                               vehicleID: "id2"),
        OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: formatter.date(from: "2024/11/13")!, 
                               distance: 1542, 
                               isMetric: false, 
                               vehicleID: "id2")
    ])
    
    return OdometerView(viewModel: viewModel)
        .environment(ActionService.shared)
}
```

mikaelacaron replied:
having the `createVehicle()` and `createReading()` functions were fine, but they should stay inside the `#Preview` macro. The functions I saw in the diff were outside of that, meaning they were just floating globally in the project, when they shouldn't be

lauriecai replied:
Leaving the functions inside results in an error on my end (which is why I took them out of preview originally):
<img width="855" alt="Screenshot 2025-03-05 at 9 46 23 AM" src="https://github.com/user-attachments/assets/c918eb3a-4ddf-4af1-8216-3adccb1113dc" />

But that led to functions floating globally - so to keep things contained inside `#Preview` I hard-coded the sample data. 

For some reason, bringing in the functions themselves results in the following error multiple times. Are you seeing the same error on your end?
```
Cannot use explicit 'return' statement in the body of result builder 'ViewBuilder'
```

mikaelacaron replied:
From what I can tell, the code was in the wrong order, this compiled for me:

and the current file on `dev` also compiles
Sorry for the delay! GitHub struggles with notifications sometimes, and I missed this because I've been busy
@lauriecai 

```swift
#Preview {
    let firstCar = createVehicle(id: "id1", name: "My 1st car")
    let secondCar = createVehicle(id: "id2", name: "2nd Car")
    
    let firstReading = createReading(vehicleID: firstCar.id!,
                                     date: "2024/10/18",
                                     distance: 35)
    let secondReading = createReading(vehicleID: firstCar.id!,
                                      date: "2024/10/19",
                                      distance: 564)
    let thirdReading = createReading(vehicleID: firstCar.id!,
                                      date: "2024/11/23",
                                      distance: 1000)
    
    let fourthReading = createReading(vehicleID: firstCar.id!,
                                     date: "2024/11/30",
                                     distance: 1024)
    let fifthReading = createReading(vehicleID: secondCar.id!,
                                      date: "2024/10/1",
                                      distance: 1000)
    
    let sixthReading = createReading(vehicleID: secondCar.id!,
                                     date: "2024/10/13",
                                     distance: 1144)
    let seventhReading = createReading(vehicleID: secondCar.id!,
                                      date: "2024/10/15",
                                      distance: 1412)
    
    let eighthReading = createReading(vehicleID: secondCar.id!,
                                     date: "2024/11/13",
                                     distance: 1542)
    
    let viewModel = OdometerViewModel(userUID: nil, firebaseService: FirebaseService())
    
    viewModel.vehicles.append(contentsOf: [firstCar, secondCar])
    // swiftlint:disable:next line_length
    viewModel.readings.append(contentsOf: [firstReading, secondReading, thirdReading, fourthReading, fifthReading, sixthReading, seventhReading, eighthReading])
    
    return OdometerView(viewModel: viewModel)
        .environment(ActionService.shared)
    
    func createVehicle(id: String, name: String) -> Vehicle {
        Vehicle(id: id, 
                userID: nil, 
                name: name, 
                make: "", 
                model: "", 
                year: nil, 
                color: nil, 
                vin: nil, 
                licensePlateNumber: nil)
    }
    
    func createReading(vehicleID: String, date: String, distance: Int) -> OdometerReading {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy/MM/dd"
        let firstDate = formatter.date(from: date)!
        return OdometerReading(id: UUID().uuidString,
                               userID: "", 
                               date: firstDate, 
                               distance: distance, 
                               isMetric: false, 
                               vehicleID: vehicleID)
    }
}
```

lauriecai replied:
That fixed it! Thank you.

mikaelacaron commented on Basic-Car-Maintenance-Tests/Shared/Odometer/MockFirebaseService.swift:
```diff
@@ -0,0 +1,87 @@
+            isMetric: true, 
+            vehicleID: "LV0000")
+    
+    let testVehicles = [
```
`testVehicles` isn't used anywhere, should probably be deleted

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/ViewModels/OdometerViewModel.swift:
```diff
@@ -48,37 +47,17 @@ class OdometerViewModel {
             readings.remove(at: eventIndex)
         }
         
+        await firebaseService.deleteReading(reading: reading, documentId: documentId)
```
here you shouldn't need to pass reading and `documentId` because the `documentId` is in the `reading` object

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/ViewModels/OdometerViewModel.swift:
```diff
@@ -87,14 +66,8 @@ class OdometerViewModel {
-                    .collection(FirestorePath.odometerReadings(vehicleID: readingToUpdate.vehicleID).path)
-                    .document(id)
-                    .setData(from: readingToUpdate)
+                try firebaseService.updateReading(reading: reading, documentId: id, userUID: userUID)
```
same here only pass the `readingToUpdate`, don't pass the 3 variables, when only one is needed

mikaelacaron commented on Basic-Car-Maintenance/Shared/Odometer/ViewModels/OdometerViewModel.swift:
```diff
@@ -108,23 +81,7 @@ class OdometerViewModel {
-                
-                self.vehicles = vehicles
-            }
+            self.vehicles = await firebaseService.getVehicles(uid: uid)
```
this parameter should be called `userID` as opposed to just `uid`, because without context "get vehicles" it's confusing why there's a uid at all and what it's for

mikaelacaron commented on Basic-Car-Maintenance/Shared/Utilities/FirebaseServiceProtocol.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+protocol FirebaseServiceProtocol {
+    func addReading(_ reading: OdometerReading) throws
+    func deleteReading(reading: OdometerReading, documentId: String) async
+    func getReadings(userUID: String) async -> [OdometerReading]
+    func updateReading(reading: OdometerReading, documentId: String, userUID: String) throws
+    func getVehicles(uid: String) async -> [Vehicle]
+}
```
sometimes an underscore `_` is used, and sometimes it isn't in the function definitions, can you make them consistent

mikaelacaron commented on Basic-Car-Maintenance-Tests/Shared/Odometer/MockFirebaseService.swift:
```diff
@@ -0,0 +1,87 @@
+                        vehicleID: "SF2222"),
+    ]
+    
+    var vehicles: [Vehicle] = []
```
the actual `FirebaseService` doesn't hold this state, why should the `MockFirebaseService` have state?

mikaelacaron commented on Basic-Car-Maintenance-Tests/Shared/Odometer/OdometerViewModelTests.swift:
```diff
@@ -0,0 +1,99 @@
+struct OdometerViewModelTests {
+
+    @Test func viewModelInitialState() throws {
+        let userUID = "testUser1"
```
Make an actual UUID here, as opposed to "testUser1" to make it more realistic

mikaelacaron commented on Basic-Car-Maintenance-Tests/Shared/Odometer/OdometerViewModelTests.swift:
```diff
@@ -0,0 +1,99 @@
+    @Test func addReading() {
+        let userUID = "testUser1"
+        let firebaseService = MockFirebaseService()
+        let viewModel = OdometerViewModel(userUID: userUID, firebaseService: firebaseService)
```
these 3 lines could be made into a function called `makeOdometerViewModel` because they're used in every test, and then each test calls that function

mikaelacaron commented on Basic-Car-Maintenance-Tests/Shared/Odometer/OdometerViewModelTests.swift:
```diff
@@ -0,0 +1,99 @@
+        
+        try? viewModel.addReading(newReading)
+        
+        #expect(firebaseService.readings.contains(newReading) == true, "New reading should be added to database")
```
right now the `firebaseService.readings` shouldn't be tested for holding the state the `odometerViewModel.readings` should. this is consistently done throughout these tests

let me know if you think otherwise, and why. but I believe the wrong thing is being tested for what's holding the state

lauriecai replied:
I did some more research and was able to incorporate tests that directly interact with the emulator. Check it out and let me know if you have additional feedback. Have requested another review.

mikaelacaron reviewed (changes requested):
Great work to start this! I've left several comments some about style, but some about how you implemented the unit tests.

If you have any questions, you can comment directly on that thread, otherwise click the resolve button once you've addressed my comment. Once you fix everything, please click the re-review button

<img width="345" alt="Re-review button" src="https://github.com/user-attachments/assets/3bdb4c27-530f-49b6-9c4c-c72e39dc12ad" />

lauriecai committed:
Merge branch 'dev' into 379-odometer-unit-tests

lauriecai committed:
Rename reading to odometerReading

lauriecai committed:
Remove singleton pattern from FirebaseService

lauriecai committed:
Remove redundant documentId argument

lauriecai committed:
Remove redundant userId argument

lauriecai committed:
Rename uid to userId

lauriecai committed:
Consolidate userId instances into userUID

lauriecai committed:
Remove function declarations in #Preview

lauriecai committed:
Remove unnecessary MockFirebaseService

lauriecai committed:
Add Equatable conformance to enable comparison

lauriecai committed:
Add test coverage for OdometerViewModel

lauriecai committed:
Fix Preview issue on OdometerView

mikaelacaron committed:
Fix merge conflicts

mikaelacaron committed:
Merge branch 'dev' into 379-odometer-unit-tests

mikaelacaron commented:
Hey @lauriecai ! I appreciate all the work you did on this PR forever ago! Hope you're doing well. I'm updating it to get it merged (just so you know, if you see any notifications about it)

mikaelacaron committed:
Cleanup

mikaelacaron committed:
Disable for now

disable because the CI needs to be properly set up to run the Firebase emulator
